### ISSUE TITLE
ShyLU/Tacho:  Can we disable Tacho by default if RDC is off

### ISSUE THREAD

**Issue by csiefer2** [trilinos/Trilinos#2580]
Most of Trilinos won't link with RDC, so it would be nice if Tacho was off by default if RDC was off (rather than issuing an error).

CMake Error at packages/shylu/shylu_node/tacho/CMakeLists.txt:8 (MESSAGE):
  ShyLu/Tacho requires CUDA relocatable device code to be enabled if CUDA is
  enabled.  Set: Kokkos_ENABLE_Cuda_Relocatable_Device_Code=ON

@jhux2

**Comment by kyungjoo-kim** [comment]
What about the other way ? @crtrott  Can we set relocatable device code is ON by default when Kokkos CUDA is enabled ? Does it have any side effect ?

**Comment by csiefer2** [comment]
@kyungjoo-kim The side effect is most of Trilinos being unable to link...

**Comment by kyungjoo-kim** [comment]
@csiefer2 

What do you mean by being unable to link with relocatable device code ? It is a compiler flag and if it does not need for the relocatable device code, is should compile. Am I missing something ? 

BTW, the cmake error pops up because Tacho is enabled. If you don't use the package, then you can simple disable the package.

**Comment by csiefer2** [comment]
@kyungjoo-kim 

1) The tests/examples for a number of packages will not link correctly if Trilinos is built with RDC on.  This is due to some very specific requirements of the linker's interactions with RDC>

2) Correct.  Unfortunately is is on by default.  Which is what we were asking to change :)

**Comment by kyungjoo-kim** [comment]
@csiefer2 Okay.

I need some cmake help. How can I distinguish the following two intentions of users ?
case 1) someone sets Tacho on and set RDC off accidently ? In this case, the compilation error should report.
case 2) Tacho is on by default and RDC is off. In this case, we want to disable Tacho is off.

**Comment by csiefer2** [comment]
@bartlettroscoe Is there a way to distinguish between "on by default" and "on because I specifically asked for it" ?

**Comment by srajama1** [comment]
@csiefer2 : There is no good way to specify a compiler flag as a dependency in Dependencies.cmake in TriBITS. That is the reason we have to resort to this configure time checks and errors. 

@kyungjoo-kim : Setting RDC always on is not a possibility now. We would like to go there, but it appears lot of code doesn't work with it.  I don't think there is a way to do what you asked in 2) above now, unless @bartlettroscoe changes TriBITS logic in some way.

**Comment by kyungjoo-kim** [comment]
Lets disable Tacho when RDC is off and returns cmake warning instead of error.

**Comment by mhoemmen** [comment]
FYI @micahahoward would like to be able to build with RDC OFF for now, though would like to relax that requirement.

**Comment by kyungjoo-kim** [comment]
Oh... that commit is not correct one. I tried on my workstation and pushed it to my local branch to test on white. It does not work that way.

 @bartlettroscoe How can I disable Trilinos_ENABLE_ShyLU_NodeTacho when Kokkos_ENABLE_Cuda is on and Kokkos_ENABLE_Cuda_Relocatable_Device_Code is off ?

**Comment by kyungjoo-kim** [comment]
@csiefer2 , would you please check if your code compiles without a problem using the PR branch mentioned above ?

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe How can I disable Trilinos_ENABLE_ShyLU_NodeTacho when Kokkos_ENABLE_Cuda is on and Kokkos_ENABLE_Cuda_Relocatable_Device_Code is off ?

@kyungjoo-kim,

As per:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-check-for-and-tweak-tribits-enable-cache-variables

you would add special logic to the file:

* Trilinos/cmake/ProjectDependenciesSetup.cmake

If you want, I can create a branch and PR that does this and test to make sure it does what it should.

If ShyLU_NodeTacho can't be used if RDC is  disabled, then we should use TriBITS logic and just disable the entire subpackage.  That will have the advantage of disabling downstream package as per:

* https://tribits.org/doc/TribitsDevelopersGuide.html#se-package-disable-triggers-auto-disables-of-downstream-dependencies

That is better than just disabling the tests as in the commit bdb742e.

Any objection to me creating this branch and PR?

**Comment by kyungjoo-kim** [comment]
@bartlettroscoe Thank you for the clarification. I would appreciate if you can provide Tribit magic and disable Tacho when RDC is disabled. That would be much better solution.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe Thank you for the clarification. I would appreciate if you can provide Tribit magic and disable Tacho when RDC is disabled. That would be much better solution.

@kyungjoo-kim, okay.  I will reopen this issue and post the PR shortly.

**Comment by bartlettroscoe** [comment]
I created PR #2586.  @kyungjoo-kim, can you please review that PR?

**Comment by bartlettroscoe** [comment]
I think this was addressed by the merge of PR #2586, right?  Can we close this as complete?

**Comment by bartlettroscoe** [comment]
NOTE: As some evidence this is working, ShyLU_NodeTacho is shown as disabled in the full PT CUDA Debug build shown at:

* https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&parentid=3455102

where the configure output at:

* https://testing-vm.sandia.gov/cdash/viewConfigure.php?buildid=3455141

shows:

```
-- NOTE: Setting Trilinos_ENABLE_ShyLU_NodeTacho=OFF by default since TPL_ENABLE_CUDA='ON' AND Kokkos_ENABLE_Cuda_Relocatable_Device_Code='OFF'!
-- NOTE: To allow the enable of ShyLU_NodeTacho, please set Kokkos_ENABLE_Cuda_Relocatable_Device_Code=ON.

...

Final set of non-enabled SE packages:  ... ShyLU_NodeTacho ... 35
```

Given this evidence, I am going to close this as complete.

If someone objects, please just reopen.
